File: packages/autonomo-python/autonomo/state.py
"""
State Manager - Tracks and reports application state

Collects state from multiple sources into a unified snapshot
that the AI can use to understand the application.
"""

from dataclasses import dataclass, field
from typing import Callable, Optional, Dict, Any, List
import time

from .registry import registry, ElementInfo
from .actions import custom_actions


@dataclass
class UserContext:
    """User context information"""
    id: Optional[str] = None
    email: Optional[str] = None
    role: Optional[str] = None
    extra: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        result = {}
        if self.id is not None:
            result["id"] = self.id
        if self.email is not None:
            result["email"] = self.email
        if self.role is not None:
            result["role"] = self.role
        result.update(self.extra)
        return result


@dataclass
class NetworkRequest:
    """Network request information"""
    method: str
    url: str
    status: Optional[int] = None
    duration: Optional[int] = None
    error: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        result = {
            "method": self.method,
            "url": self.url,
        }
        if self.status is not None:
            result["status"] = self.status
        if self.duration is not None:
            result["duration"] = self.duration
        if self.error is not None:
            result["error"] = self.error
        return result


@dataclass
class AppState:
    """Complete application state snapshot"""
    screen: str
    timestamp: int
    user: Optional[UserContext]
    elements: List[ElementInfo]
    custom_actions: List[str]
    data: Optional[Dict[str, Any]]
    errors: List[str]
    logs: List[str]
    render_errors: List[str]
    network: Optional[List[NetworkRequest]]

    def to_dict(self) -> Dict[str, Any]:
        result = {
            "screen": self.screen,
            "timestamp": self.timestamp,
            "elements": [e.to_dict() for e in self.elements],
            "customActions": self.custom_actions,
            "errors": self.errors,
            "logs": self.logs,
            "renderErrors": self.render_errors,
        }
        if self.user is not None:
            result["user"] = self.user.to_dict()
        if self.data is not None:
            result["data"] = self.data
        if self.network is not None:
            result["network"] = [n.to_dict() for n in self.network]
        return result


class StateManager:
    """Singleton state manager"""
    
    _instance = None
    
    MAX_ERRORS = 50
    MAX_LOGS = 100
    MAX_NETWORK = 50
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._screen = "unknown"
            cls._instance._user: Optional[UserContext] = None
            cls._instance._data: Dict[str, Any] = {}
            cls._instance._errors: List[str] = []
            cls._instance._logs: List[str] = []
            cls._instance._render_errors: List[str] = []
            cls._instance._network: List[NetworkRequest] = []
            cls._instance._listeners: List[Callable[[AppState], None]] = []
            
            # Forward registry/action changes
            registry.on_change(lambda: cls._instance.notify_change())
            custom_actions.on_change(lambda: cls._instance.notify_change())
        return cls._instance
    
    def set_screen(self, screen: str) -> None:
        """Set current screen/route"""
        self._screen = screen
        self.notify_change()
    
    def get_screen(self) -> str:
        """Get current screen"""
        return self._screen
    
    def set_user(self, user: Optional[UserContext]) -> None:
        """Set user context"""
        self._user = user
        self.notify_change()
    
    def set_data(self, data: Dict[str, Any]) -> None:
        """Set application data"""
        self._data = data
        self.notify_change()
    
    def merge_data(self, data: Dict[str, Any]) -> None:
        """Merge data into existing"""
        self._data.update(data)
        self.notify_change()
    
    def add_error(self, error: str) -> None:
        """Add an error"""
        self._errors.append(error)
        if len(self._errors) > self.MAX_ERRORS:
            self._errors = self._errors[-self.MAX_ERRORS:]
        self.notify_change()
    
    def add_log(self, log: str) -> None:
        """Add a log entry"""
        self._logs.append(log)
        if len(self._logs) > self.MAX_LOGS:
            self._logs = self._logs[-self.MAX_LOGS:]
    
    def add_render_error(self, error: str) -> None:
        """Add a render error"""
        self._render_errors.append(error)
        if len(self._render_errors) > self.MAX_ERRORS:
            self._render_errors = self._render_errors[-self.MAX_ERRORS:]
        self.notify_change()
    
    def add_network_request(self, request: NetworkRequest) -> None:
        """Add a network request"""
        self._network.append(request)
        if len(self._network) > self.MAX_NETWORK:
            self._network = self._network[-self.MAX_NETWORK:]
    
    def clear_errors(self) -> None:
        """Clear errors"""
        self._errors.clear()
        self._render_errors.clear()
        self.notify_change()
    
    def clear_logs(self) -> None:
        """Clear logs"""
        self._logs.clear()
    
    def clear_network(self) -> None:
        """Clear network history"""
        self._network.clear()
    
    def get_state(self) -> AppState:
        """Get current state snapshot"""
        return AppState(
            screen=self._screen,
            timestamp=int(time.time() * 1000),
            user=self._user,
            elements=registry.get_all(),
            custom_actions=custom_actions.list(),
            data=self._data if self._data else None,
            errors=list(self._errors),
            logs=list(self._logs),
            render_errors=list(self._render_errors),
            network=list(self._network) if self._network else None,
        )
    
    def on_change(self, listener: Callable[["AppState"], None]) -> Callable[[], None]:
        """Subscribe to state changes"""
        self._listeners.append(listener)
        return lambda: self._listeners.remove(listener)
    
    def notify_change(self) -> None:
        """Trigger a state update notification"""
        state = self.get_state()
        for listener in self._listeners:
            listener(state)


# Singleton instance
state = StateManager()

File: packages/autonomo_flutter/lib/src/state.dart
/// State Manager - Tracks and reports application state
///
/// Collects state from multiple sources into a unified snapshot
/// that the AI can use to understand the application.

import 'registry.dart';
import 'actions.dart';
import 'instance.dart';

/// User context information
class UserContext {
  final String? id;
  final String? email;
  final String? role;
  final Map<String, dynamic> extra;

  const UserContext({
    this.id,
    this.email,
    this.role,
    this.extra = const {},
  });

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        if (email != null) 'email': email,
        if (role != null) 'role': role,
        ...extra,
      };
}

/// Network request information
class NetworkRequest {
  final String method;
  final String url;
  final int? status;
  final int? duration;
  final String? error;

  const NetworkRequest({
    required this.method,
    required this.url,
    this.status,
    this.duration,
    this.error,
  });

  Map<String, dynamic> toJson() => {
        'method': method,
        'url': url,
        if (status != null) 'status': status,
        if (duration != null) 'duration': duration,
        if (error != null) 'error': error,
      };
}

/// Suggested action in a flow
class SuggestedAction {
  /// Action to perform: press, fillIn, navigate, custom
  final String action;
  /// Target element ID or route
  final String target;
  /// Optional value for fillIn or custom actions
  final String? value;
  /// Human-readable description of this step
  final String? description;

  const SuggestedAction({
    required this.action,
    required this.target,
    this.value,
    this.description,
  });

  Map<String, dynamic> toJson() => {
        'action': action,
        'target': target,
        if (value != null) 'value': value,
        if (description != null) 'description': description,
      };
}

/// Complete application state snapshot
class AppState {
  final String screen;
  final int timestamp;
  final InstanceInfo? instance;
  final UserContext? user;
  /// Screen-level hint for AI agents.
  /// Provides context about what this screen is for and how to use it.
  final String? screenHint;
  /// Suggested flow of actions for this screen.
  /// Helps AI understand the typical workflow.
  final List<SuggestedAction>? suggestedFlow;
  final List<ElementInfo> elements;
  final List<String> customActions;
  final Map<String, dynamic>? data;
  final List<String> errors;
  final List<String> logs;
  final List<String> renderErrors;
  final List<NetworkRequest>? network;

  const AppState({
    required this.screen,
    required this.timestamp,
    this.instance,
    this.user,
    this.screenHint,
    this.suggestedFlow,
    required this.elements,
    required this.customActions,
    this.data,
    required this.errors,
    required this.logs,
    required this.renderErrors,
    this.network,
  });

  Map<String, dynamic> toJson() => {
        'screen': screen,
        'timestamp': timestamp,
        if (instance != null) 'instance': instance!.toJson(),
        if (user != null) 'user': user!.toJson(),
        if (screenHint != null) 'screenHint': screenHint,
        if (suggestedFlow != null)
          'suggestedFlow': suggestedFlow!.map((s) => s.toJson()).toList(),
        'elements': elements.map((e) => e.toJson()).toList(),
        'customActions': customActions,
        if (data != null) 'data': data,
        'errors': errors,
        'logs': logs,
        'renderErrors': renderErrors,
        if (network != null)
          'network': network!.map((n) => n.toJson()).toList(),
      };
}

typedef StateChangeListener = void Function(AppState state);

/// Singleton state manager
class StateManager {
  static final StateManager _instance = StateManager._internal();
  factory StateManager() => _instance;
  StateManager._internal() {
    // Forward registry changes to state listeners
    registry.onChange(() => notifyChange());
    customActions.onChange(() => notifyChange());
  }

  String _screen = 'unknown';
  String? _screenHint;
  List<SuggestedAction>? _suggestedFlow;
  UserContext? _user;
  Map<String, dynamic> _data = {};
  final List<String> _errors = [];
  final List<String> _logs = [];
  final List<String> _renderErrors = [];
  final List<NetworkRequest> _network = [];
  final Set<StateChangeListener> _listeners = {};

  static const int _maxErrors = 50;
  static const int _maxLogs = 100;
  static const int _maxNetwork = 50;

  /// Set current screen/route
  /// Clears screen-specific hints when screen changes
  void setScreen(String screen) {
    _screen = screen;
    _screenHint = null;
    _suggestedFlow = null;
    notifyChange();
  }

  /// Get current screen
  String getScreen() => _screen;

  /// Set screen-level hint for AI agents.
  /// Provides context about what this screen is for and how to use it.
  void setScreenHint(String? hint) {
    _screenHint = hint;
    notifyChange();
  }

  /// Set suggested flow of actions for the current screen.
  /// Helps AI understand the typical workflow.
  void setSuggestedFlow(List<SuggestedAction>? flow) {
    _suggestedFlow = flow;
    notifyChange();
  }

  /// Set screen with optional hint and suggested flow.
  /// Convenience method to set all screen context at once.
  void setScreenContext(String screen, {String? hint, List<SuggestedAction>? flow}) {
    _screen = screen;
    _screenHint = hint;
    _suggestedFlow = flow;
    notifyChange();
  }

  /// Set user context
  void setUser(UserContext? user) {
    _user = user;
    notifyChange();
  }

  /// Set application data
  void setData(Map<String, dynamic> data) {
    _data = data;
    notifyChange();
  }

  /// Merge data into existing
  void mergeData(Map<String, dynamic> data) {
    _data = {..._data, ...data};
    notifyChange();
  }

  /// Add an error
  void addError(String error) {
    _errors.add(error);
    if (_errors.length > _maxErrors) {
      _errors.removeRange(0, _errors.length - _maxErrors);
    }
    notifyChange();
  }

  /// Add a log entry
  void addLog(String log) {
    _logs.add(log);
    if (_logs.length > _maxLogs) {
      _logs.removeRange(0, _logs.length - _maxLogs);
    }
  }

  /// Add a render error
  void addRenderError(String error) {
    _renderErrors.add(error);
    if (_renderErrors.length > _maxErrors) {
      _renderErrors.removeRange(0, _renderErrors.length - _maxErrors);
    }
    notifyChange();
  }

  /// Add a network request
  void addNetworkRequest(NetworkRequest request) {
    _network.add(request);
    if (_network.length > _maxNetwork) {
      _network.removeRange(0, _network.length - _maxNetwork);
    }
  }

  /// Clear errors
  void clearErrors() {
    _errors.clear();
    _renderErrors.clear();
    notifyChange();
  }

  /// Clear logs
  void clearLogs() {
    _logs.clear();
  }

  /// Clear network history
  void clearNetwork() {
    _network.clear();
  }

  /// Get current state snapshot
  AppState getState() {
    return AppState(
      screen: _screen,
      timestamp: DateTime.now().millisecondsSinceEpoch,
      instance: getInstance(),
      user: _user,
      screenHint: _screenHint,
      suggestedFlow: _suggestedFlow,
      elements: registry.getAll(),
      customActions: customActions.list(),
      data: _data.isNotEmpty ? _data : null,
      errors: List.unmodifiable(_errors),
      logs: List.unmodifiable(_logs),
      renderErrors: List.unmodifiable(_renderErrors),
      network: _network.isNotEmpty ? List.unmodifiable(_network) : null,
    );
  }

  /// Subscribe to state changes
  void Function() onChange(StateChangeListener listener) {
    _listeners.add(listener);
    return () => _listeners.remove(listener);
  }

  /// Trigger a state update notification
  void notifyChange() {
    final state = getState();
    for (final listener in _listeners) {
      listener(state);
    }
  }
}

/// Global state manager instance
final state = StateManager();
